qubes/ext/block.py

```diff
@@ -283,6 +458,23 @@ def on_device_pre_attached_block(self, vm, event, device, options):
             vm.app.env.get_template('libvirt/devices/block.xml').render(
                 device=device, vm=vm, options=options))
 
+    @qubes.ext.handler('domain-start')
+    async def on_domain_start(self, vm, _event, **_kwargs):
+        # pylint: disable=unused-argument
+        for assignment in vm.devices['block'].get_assigned_devices():
+            asyncio.ensure_future(self._attach_and_notify(
+                vm, assignment.device, assignment.options))
```

Comment:
Block devices attached at domain start can be specified in libvirt xml, in fact you already adjust that part.

Reply:
Can you elaborate?

Reply:
Block devices that are assigned when VM is starting are attached already at https://github.com/QubesOS/qubes-core-admin/pull/579/files#diff-bd38d7fc85084f576ddb6d1d4702f72158e0e30d5e1deb15cf3d49c28c0eb82eR152-R156. So, it looks like you attach them again at `domain-start` event here.

Reply:
This is still the case.

Reply:
Block devices attached at domain startup should be specified in libvirt xml (linked above), instead of attached after domain startup. This is especially important for "required" devices, as one may need them during system startup (for example some standalone qube may have it listed in /etc/fstab and startup will fail without it).